perfkitbenchmarker/pkb.py

```diff
@@ -195,6 +195,51 @@
 events.initialization_complete.connect(traces.RegisterAll)
 
 
+def _CreateBenchmarkRunList():
+  """Create a list of configs and states for each benchmark run to be scheduled.
+
+  Returns:
+    list of (args, run_status_list) pairs. Contains one pair per benchmark run
```

Comment:
I agree that things are a bit complicated at the moment. I see some duplication in fields of the args and the run_status_list, and they could be worked out.

In the case of staged runs, the status actually represents whether the desired stage was run and whether it was successful for each benchmark. So making it part of the BenchmarkSpec may not make sense, since it's a state that has more to do with the current PKB process, and not the overall benchmark run (provision-prepare-run-cleanup-teardown).

The only thing missing from a partial is the collector argument, which currently is only created once and shared for all benchmarks. It would need to be passed to this function, which somehow seems unclean to me.

Perhaps a stronger type definition or two might help. What we essentially have at this point is:
1) Information needed by RunBenchmark.
2) Information needed by benchmark_status.CreateSummary.

The history behind building the lists ahead of time is also something that is tied to our now-removed --parallelism flag, so there may not even be a need for it.

I guess what I'm leaning toward at this time is something like a map of uid to (module, config, status), and adapting the later loop in RunBenchmarks and CreateSummary to use the map. But it's a somewhat sizable change from the current behavior, so I'd prefer to think a bit more and then make a separate PR.